### ISSUE TITLE
Versioning scheme / release cycle?

### ISSUE THREAD

**Issue by veghp** [biopython/biopython#3265]
First I would like to thank you for writing and maintaining Biopython, we find it very useful.

I have a few questions about the release procedure. I'm wondering what versioning scheme is used and what rules are followed regarding releases? For example, when is a release made?

(I tried looking for this info in the [Contributing doc](https://biopython.org/wiki/Contributing), or [rst](https://github.com/biopython/biopython/blob/master/CONTRIBUTING.rst) or here: [Building a release](https://biopython.org/wiki/Building_a_release).)
I infer from a comment in https://github.com/biopython/biopython/issues/3243 that PEP440 is followed for the format of the version string. 
 
 
Related to this, are all API changes logged in [NEWS.rst](https://github.com/biopython/biopython/blob/master/NEWS.rst)?

This would be helpful to users. For example, PR https://github.com/biopython/biopython/pull/3048, in addition to removing alphabets, the renamed the `format` parameter of the Bio.motifs.parse function to `fmt` (cf. [documentation](https://biopython.org/DIST/docs/api/Bio.motifs-pysrc.html#parse)).

**Comment by peterjc** [comment]
Releases are roughly every three to four months, depending on time availability although improved automation has made this easier.

The version number is simply incremental, no semantic versioning here. We try to preserve backward compatibility but small breaks do happen. See also DEPRECATED.rst

Major changes are logged in the NEWS.rst file.

In hindsight switching that argument name from "format" (no a reserved word in Python) to "fmt" in #3048 was perhaps unwise (CC @mdehoon), but it probably isn't worth changing it back. In that case positional arguments would work for calling the old or new function.

https://biopython.org/DIST/docs/api/ is for Biopython 1.75, we ought to have that redirecting to https://biopython.org/docs/1.78/api/ or https://biopython.org/docs/dev/api/ instead (see #2905; see also #2906 on version switching).

**Comment by mdehoon** [comment]
It would be a good project for an enterprising Biopythoneer to change all variables with the name "format" to "fmt" or "format_spec". Currently there are 2539 lines in Biopython that contain the word "format". Using the built-in function "format" as a variable name is asking for trouble.
Best,-Michiel

    On Wednesday, September 9, 2020, 11:04:09 PM GMT+9, Peter Cock <notifications@github.com> wrote:  
 
 


Releases are roughly every three to four months, depending on time availability although improved automation has made this easier.

The version number is simply incremental, no semantic versioning here. We try to preserve backward compatibility but small breaks do happen. See also DEPRECATED.rst

Major changes are logged in the NEWS.rst file.

In hindsight switching that argument name from "format" (no a reserved word in Python) to "fmt" in #3048 was perhaps unwise (CC @mdehoon), but it probably isn't worth changing it back. In that case positional arguments would work for calling the old or new function.

https://biopython.org/DIST/docs/api/ is for Biopython 1.75, we ought to have that redirecting to https://biopython.org/docs/1.78/api/ or https://biopython.org/docs/dev/api/ instead (see #2905; see also #2906 on version switching).

**Comment by peterjc** [comment]
Changing ``format`` on internal code is fine, changing it in argument names or attributes is a breaking API change which is not fine.

**Comment by veghp** [comment]
Thank you for the quick replies. Yes I used positional arguments, which solved the problem. Thank you

**Comment by mdehoon** [comment]
> Changing `format` on internal code is fine, changing it in argument names or attributes is a breaking API change which is not fine.

Why not? With Python itself, there are breaking API changes in every minor release.

The alternative would be to live forever with our mistakes from the past.

**Comment by mdehoon** [comment]
We should also remove `format` as a member function, and use `__format__` instead.

**Comment by peterjc** [comment]
It is still a draft, but the introduction in NumPy's NEP on backwards compatibility resonates with me: https://numpy.org/neps/nep-0023-backwards-compatibility.html

Changing existing function arguments or attributes to avoid built-in functions names like ``format`` (``type``, ``id``) to me is a cosmetic improvement and fails their "*Aim not to break users’ code unnecessarily*" test.

#2493 already deprecated the MSA object's ``.format`` method and others in favour of using ``__format__`` (released in May 2020 in Biopython 1.77, so not long till they can be removed):
https://github.com/biopython/biopython/blob/biopython-178/Bio/Align/__init__.py#L298

I have no objection to deprecating and later removing the ``SeqRecord`` object's ``.format`` method in the same way:
https://github.com/biopython/biopython/blob/biopython-178/Bio/SeqRecord.py#L688

**Comment by mdehoon** [comment]
Quoting from their NumPy's NEP:

> When assessing the costs, keep in mind that most users do not read the mailing list, do not look at deprecation warnings, and sometimes wait more than one or two years before upgrading from their old version. And that NumPy has many hundreds of thousands or even a couple of million users, so “no one will do or use this” is very likely incorrect.

- I would agree that most users do not look at deprecation warnings. They are pretty much useless.
- In terms of user base, we are not quite in the same ballpark as NumPy.
- Changing code to avoid built-in function names is not just a cosmetic change; it is done to avoid bugs. I have run into such bugs myself, where I thought I was checking a variable named `format`, and later realized it was evaluating the `format` built-in function instead.

**Comment by veghp** [comment]
Dear Michiel, I appreciate you developing and maintaining the motifs module. Perhaps I can add to the discussion from one user's point of view. As for implementing changes, I agree with @peterjc and I liked the future-warning approach for alphabets with both old and new versions being supported for a while. Reliability is key to us as Biopython is used by many [packages](https://edinburgh-genome-foundry.github.io/) and [apps](https://cuba.genomefoundry.org/) that are critical to our lab operations. For example, motifs are used by [DNA Chisel](https://github.com/Edinburgh-Genome-Foundry/dnachisel), which in turn is used by Benchling, which has tens of thousands of customers.

Additionally, perhaps these modifications should be made in separate PRs or at least in separate commits, as commits ought to contain a coherent set of changes which also makes work easier during debugging. Alternatively, perhaps a rule can be established that all API changes must be logged in NEWS.rst.

**Comment by mdehoon** [comment]
@veghp The problem with the current emphasis on backward compatibility is that improvements to Biopython have been extremely slow. Just removing alphabets took 15 years (this is not a typo). There are still a number of fundamental flaws in Biopython that at the current rate would take beyond my age of retirement to fix.

Perhaps this is a good time to split off a 2.0 branch for users who want to see faster improvements and are willing to pay the price of the code being less stable. Once the 2.0 branch stabilizes, users can decide to switch to 2.0 at the time of their choosing.

**Comment by peterjc** [comment]
Perhaps so, and people used to semantic versioning would recognise and expect API breakage going from v1.x to v2.x (or v2.x.y). #3188 would be a good candidate. Perhaps type checking too (#3165) but that ought to be possible on v1.x as it isn't a breaking change.

**Comment by mdehoon** [comment]
@peterjc  For a 2.0 branch, it may be a good idea to write the documentation (i.e. the Tutorial & Cookbook) first, before doing any coding, to reach a consensus on the fundamentals (in particular for the Seq object and the alignment object).

**Comment by peterjc** [comment]
Documentation (and test?) driven development? Sounds sensible - certainly I would agree with starting with a vision and plan, and not just a free for all.

**Comment by JoaoRodrigues** [comment]
Re: 2.0 sounds like a good plan indeed.

As for version changes breaking APIs, I agree that we should document any changes in NEWS. I also agree that maintaining backwards compatibility is sometimes a nuisance when we want to improve things, but I see the point raised by @veghp that Biopython is included as a dependency is other popular packages. That said, if we do use `format` as a function in places, we should rename it as soon as possible to avoid problems, or at the very least, confusion.

Finally, I actually appreciate the DeprecationWarnings and I think they are very useful. If users don't read them, it's not our problem really is it?

**Comment by chris-rands** [comment]
+1 for the biopython v 2.0 idea. Essential bug fixes could still be back-ported to v 1.x. The only negative I see is that it's a lot of work to setup and maintain...

Version 2 could include the exciting ideas in #3188 (unification of `Seq` and `SeqRecord`, use of the buffer protocol etc.).

The basic directory structures and modules could also be changed and PEP compliant, e.g. instead of `import Bio` I like `import biopython as bp`. This would fix issues like #2363 and IMO be more intuitive for users

**Comment by peterjc** [comment]
Yes, I think I've said this before but v2 would be a chance to make the module names lower case PEP8 compliant. Switching the top level from ``Bio/`` to ``biopython/`` would also allow parallel installation alongside v1 which has other benefits (e.g. large dependency stack with some tools wanting v1 and others v2).

**Comment by mdehoon** [comment]
As PR #3188 won't go into the Biopython 1.xx series, I closed the PR and renamed the branch behind it to https://github.com/mdehoon/biopython/tree/my_biopython_2.0 . Purpose of this branch is just to collect and try out some ideas for Biopython 2.0.

**Comment by mdehoon** [comment]
> Yes, I think I've said this before but v2 would be a chance to make the module names lower case PEP8 compliant.

Yes we should do that.

> Switching the top level from Bio/ to biopython/ would also allow parallel installation alongside v1 which has other benefits (e.g. large dependency stack with some tools wanting v1 and others v2).

I would still prefer `import bio` as it's shorter, and for name recognition. Being able to install v1 and v2 in parallel is a temporary goal which we may not care about in two years time, while the module names will stay with us for the next 20 years, if not more. Note that Mac and Unix distinguish `import bio` and `import Bio`.

**Comment by peterjc** [comment]
We've had this conversation before, but on Windows and many Macs ``bio/`` versus ``Bio/`` breaks badly. That would be a bad idea - see e.g. #2363

**Comment by mdehoon** [comment]
@peterjc  After we release a biopython-2.0 with `biopython` as the module name, some other package may start using `bio`. What will we do then? We'd still have the `bio` vs `Bio` confusion, but we won't have any control over it.

**Comment by peterjc** [comment]
Whatever we pick, nothing stops other packages picking the same folder names - we have no control now. At least we have reserved "biopython" (case does not matter) on PyPI, and as discussed on #2363 could now claim "bio" as well.

**Comment by chris-rands** [comment]
I think naively when people run `pip install biopython` they expect to be able to `import biopython`, this follows the convention of many python packages (e.g. numpy, pandas, matplotlib, etc.). If you want a short form name why not encourage `import biopython as bp` or `import biopython as bio`? Think of `import numpy as np` and `import pandas as pd`

**Comment by mdehoon** [comment]
@chris-rands  
> If you want a short form name why not encourage `import biopython as bp` or `import biopython as bio`

It wouldn't help much; the `Bio` module contains just a few definitions of errors and warnings, but not much of interest to the user. Most imports by users are of the form
```python
>>> from Bio import SeqIO
>>> from Bio import Entrez
>>> from Bio.Align import PairwiseAligner
>>> from Bio import Cluster
```
These would become
```python
>>> from biopython import seqIO
>>> from biopython import entrez
>>> from biopython.align import PairwiseAligner
>>> from biopython import cluster
```
and `import biopython as bp` would not make these imports any shorter.

**Comment by chris-rands** [comment]
@mdehoon True, the biopython setup would be more like matplotlib, e.g. 
```
from matplotlib import pyplot as pl
from matplotlib import rcParams, ticker, gridspec, axes
from matplotlib.axes import Axes
```

Of course it is longer to write `biopython` but you only do it once per import per module.

There are [140 github repos ](https://github.com/ylathouris/bio/network/dependents) with `bio`/`Bio` listed in their `requirements.txt` (this number is growing quite rapidly, it was 126 on 2 August), I think all of these should be `biopython`. In the long term this confusion should disappear if biopython v2 adopts `import biopython`

**Comment by chris-rands** [comment]
[PEP 423](https://www.python.org/dev/peps/pep-0423/) has some recommendations on package naming, although note the PEP is not accepted but has status "deferred". It includes [Use a single name](https://www.python.org/dev/peps/pep-0423/#use-a-single-name):

> Distribute only one package (or only one module) per project, and use package (or module) name as project name.
> 
> It avoids possible confusion between project name and distributed package or module name.
> 
> It makes the name consistent.
> 
> It is explicit: when one sees project name, he guesses package/module name, and vice versa.
> 
> It also limits implicit clashes between package/module names. By using a single name, when you register a project name to PyPI [11], you also perform a basic package/module name availability verification.
> 
> As an example, pipeline [13], python-pipeline [14] and django-pipeline [15] all distribute a package or module called "pipeline". So installing two of them leads to errors. This issue wouldn't have occurred if these distributions used a single name.

The advice suggests (as I interpret it) that the top level directory for biopython could be `biopython/` rather than `bio/` or `Bio/`. It also suggests that it might be worth making BioSQL a separate package from biopython for v2. I have never used BioSQL myself so don't know if that is a good idea?

**Comment by peterjc** [comment]
If we kept BioSQL in v2, it could easily live as a subpackage under the main namespace - I see no strong reason to keep it top level.

**Comment by MarkusPiotrowski** [comment]
> I think naively when people run `pip install biopython` they expect to be able to `import biopython`, this follows the convention of many python packages (e.g. numpy, pandas, matplotlib, etc.).

We have even seen it vice versa. People see `import Bio` (in a tutorial etc) and did do `pip install Bio` which lead to most bio/Bio problems
